### PR TITLE
Save and load training data in FP32

### LINKED ISSUE #433: Accuracy wildly wrong when loading numpy32 data
For some reason, loading numpy32 data gives wildly wrong results for LDOS integration. This seems like a bug, since NN internally use FP32.

### PR DESCRIPTION
**Pull request by RandomDefaultUser**
Since pytorch operates with FP32 anyway, it would be a good idea to have the data saved in FP32 too. This seemed like a problem since LDOS integration seemed to go wrong when performed with FP32 data, but this was an oversight in the usage of the `np.sum` function. This PR fixes this, closes #433 and allows for native training data conversion into FP32. FP64 conversion is supported as legacy option.

### CODE REVIEW

File: mala/descriptors/bispectrum.py [hunk 1]
```diff
@@ -175,13 +181,14 @@ def _calculate(self, atoms, outdir, grid_dimensions, **kwargs):
             # deallocates memory too quickly. This copy is more memory
             # hungry, and we might have to tackle this later on, but
             # for now it works.
-            return snap_descriptors_np.copy(), nrows_local
+            return snap_descriptors_np, nrows_local
```
Comment (elcorto) [inline mala/descriptors/bispectrum.py:184]: The comment above says that the `copy()` is needed, but this PR removes it.

File: mala/descriptors/bispectrum.py [hunk 2]
```diff
@@ -196,6 +203,6 @@ def _calculate(self, atoms, outdir, grid_dimensions, **kwargs):
             # it does not necessarily do that - so we have do go down
             # that route.
             if self.parameters.descriptors_contain_xyz:
-                return snap_descriptors_np.copy(), nx*ny*nz
+                return snap_descriptors_np, nx*ny*nz
             else:
-                return snap_descriptors_np[:, :, :, 3:].copy(), nx*ny*nz
```
Comment (elcorto) [inline mala/descriptors/bispectrum.py:201]: The comment above says that the copy() is needed, but this PR removes it.

File: mala/descriptors/bispectrum.py [hunk 3]
```diff
@@ -196,6 +203,6 @@ def _calculate(self, atoms, outdir, grid_dimensions, **kwargs):
             # it does not necessarily do that - so we have do go down
             # that route.
             if self.parameters.descriptors_contain_xyz:
-                return snap_descriptors_np.copy(), nx*ny*nz
```
[the remark this answers is not in the document]
Comment (RandomDefaultUser) [reply mala/descriptors/bispectrum.py:199]: The copy has been moved in the `lammps_util.py` file, but you are right, the comment should be adapted. Will do that.

File: mala/descriptors/bispectrum.py (hunk 1 above)
Comment (RandomDefaultUser) [reply mala/descriptors/bispectrum.py:184]: The copy has been moved in the lammps_util.py file, but you are right, the comment should be adapted. Will do that.

File: mala/descriptors/bispectrum.py (hunk 2 above)
Comment (RandomDefaultUser) [reply mala/descriptors/bispectrum.py:201]: The copy has been moved in the lammps_util.py file, but you are right, the comment should be adapted. Will do that.

**Comment by RandomDefaultUser** [comment]
> General question: would is be useful to replace the explicit `np.float32` by `ldos_dtype` or something like a `DEFAULT_NP_DATA_DTYPE`, which would be a package-wide default?

Good question. I am not sure. I personally think that FP32 is the reasonable default for every application here, because torch internally uses it always (or our torch code, that is). So I'd argue for giving the FP64 only as a legacy / very very specific option to users to avoid confusion.

File: mala/descriptors/bispectrum.py (hunk 2 above)
Comment (RandomDefaultUser) [reply mala/descriptors/bispectrum.py:201]: Fixed by 2c7cf122acf4eca51f632def689b93414de177a5

File: mala/descriptors/bispectrum.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (RandomDefaultUser) [reply mala/descriptors/bispectrum.py:199]: Fixed by 2c7cf122acf4eca51f632def689b93414de177a5

**Comment by elcorto** [comment]
> > General question: would is be useful to replace the explicit `np.float32` by `ldos_dtype` or something like a `DEFAULT_NP_DATA_DTYPE`, which would be a package-wide default?
> 
> Good question. I am not sure. I personally think that FP32 is the reasonable default for every application here, because torch internally uses it always (or our torch code, that is). So I'd argue for giving the FP64 only as a legacy / very very specific option to users to avoid confusion.

Sure. I guess I ment this more in a coding-only fashion: Replace explicit occurences of `np.float32` with something like `common.parameters.DEFAULT_NP_DATA_DTYPE` which would be equal to `np.float32`, *not* exposing this as a setting, say in `common.parameters.ParametersData`. This would be a way to change that in one place if needed. But that's not a high prio feature. Would affect those lines:

```
examples/ex11_pass_single_feature.py:    x = torch.from_numpy(inputs_array[0, 0, 0, 3:].astype(np.float32))
mala/datahandling/data_handler.py:        # (i.e. float64 to float32 conversion) saving the data this way
mala/datahandling/data_handler.py:                                                 dtype=np.float32)
mala/datahandling/data_handler.py:                                                  dtype=np.float32)
mala/datahandling/data_handler.py:                                                   dtype=np.float32)
mala/datahandling/data_handler.py:                                                    dtype=np.float32)
mala/datahandling/data_handler.py:                                             dtype=np.float32)
mala/datahandling/data_handler.py:                                              dtype=np.float32)
mala/datahandling/data_handler.py:                    tmp = tmp.astype(np.float32)
mala/datahandling/data_handler.py:                    tmp = tmp.astype(np.float32)
mala/datahandling/data_handler.py:        numpy_array = numpy_array.astype(np.float32)
mala/datahandling/lazy_load_dataset.py:        self.input_data = self.input_data.astype(np.float32)
mala/datahandling/lazy_load_dataset.py:            self.output_data = self.output_data.astype(np.float32)
mala/datahandling/lazy_load_dataset_clustered.py:        input_data = input_data.astype(np.float32)
mala/datahandling/lazy_load_dataset_clustered.py:        self.input_data = self.input_data.astype(np.float32)
mala/datahandling/lazy_load_dataset_clustered.py:            self.output_data = self.output_data.astype(np.float32)
mala/network/predictor.py:                    snap_descriptors.astype(np.float32)
mala/network/predictor.py:                snap_descriptors.astype(np.float32)
test/scaling_test.py:# the float32 converisons.
test/scaling_test.py:            data = data.astype(np.float32)
test/scaling_test.py:            data2 = data2.astype(np.float32)
test/tensor_memory_test.py:        loaded_array_raw = loaded_array_raw.astype(np.float32)
```

**Review by elcorto — COMMENTED** [review]
This can be made shorter by `some_dict.get(key, default_value)`. That returns `default_value` if `key not in some_dict`.

**Comment by RandomDefaultUser** [comment]
> This can be made shorter by `some_dict.get(key, default_value)`. That returns `default_value` if `key not in some_dict`.

Good point, I should have done that to begin with, there were also other instances where I used these clunky if-s. Fixed with c783ab2615e6a931f24b0983f3865f72f9fc48b0.

**Comment by elcorto** [comment]
In terms of saving space by using float32, #436 is maybe an even better option for the future.
